File: generated_code/optimized/runtime/cpp/4.cpp
#include <vector>
#include <cmath>
using namespace std;

float mean_absolute_deviation(vector<float> numbers) {
    const size_t n = numbers.size();
    const float* p = numbers.data();

    float sum = 0.0f;
    for (size_t i = 0; i < n; ++i) sum += p[i];

    const float avg = sum / static_cast<float>(n);

    float msum = 0.0f;
    for (size_t i = 0; i < n; ++i) msum += fabsf(p[i] - avg);

    return msum / static_cast<float>(n);
}

File: generated_code/optimized/runtime/java/4.java
import java.util.*;
import java.lang.*;

class Solution {
    public double meanAbsoluteDeviation(List<Double> numbers) {
        final int n = numbers.size();
        if (n == 0) return Double.NaN;

        double sum = 0.0;

        if (numbers instanceof RandomAccess) {
            for (int i = 0; i < n; i++) {
                sum += numbers.get(i);
            }
            final double mean = sum / n;
            double sumAbsDiff = 0.0;
            for (int i = 0; i < n; i++) {
                sumAbsDiff += Math.abs(numbers.get(i) - mean);
            }
            return sumAbsDiff / n;
        } else {
            for (Double num : numbers) {
                sum += num;
            }
            final double mean = sum / n;
            double sumAbsDiff = 0.0;
            for (Double num : numbers) {
                sumAbsDiff += Math.abs(num - mean);
            }
            return sumAbsDiff / n;
        }
    }
}
